alf/drivers/on_policy_driver.py

```diff
@@ -0,0 +1,370 @@
+            traj = from_transition(time_step,
+                                   policy_step._replace(action=action),
+                                   next_time_step)
+            for observer in self._observers:
+                observer(traj)
+
+        return next_time_step, policy_step, action
+
+    def _train_loop_body(self, counter, time_step, policy_state,
+                         training_info_ta):
+
+        next_time_step, policy_step, action = self._step(
+            time_step, policy_state)
+        action_distribution_param = get_distribution_params(policy_step.action)
+
+        training_info = TrainingInfo(
+            action_distribution=action_distribution_param,
+            action=action,
+            reward=time_step.reward,
+            discount=time_step.discount,
+            step_type=time_step.step_type,
+            info=policy_step.info)
+
+        training_info_ta = tf.nest.map_structure(
+            lambda ta, x: ta.write(counter, x), training_info_ta,
+            training_info)
+
+        counter += 1
+
+        return [counter, next_time_step, policy_step.state, training_info_ta]
+
+    def _eval_loop_body(self, time_step, policy_state):
+        next_time_step, policy_step, _ = self._step(time_step, policy_state)
+
+        return [next_time_step, policy_step.state]
+
+    def _iter(self, time_step, policy_state):
+        """One training iteration."""
+        counter = tf.zeros((), tf.int32)
+        batch_size = self.env.batch_size
+
+        def create_ta(s):
+            return tf.TensorArray(
+                dtype=s.dtype,
+                size=self._train_interval,
+                element_shape=tf.TensorShape([batch_size]).concatenate(
+                    s.shape))
+
+        training_info_ta = tf.nest.map_structure(create_ta,
+                                                 self._training_info_spec)
+
+        with tf.GradientTape() as tape:
+            [_, time_step, policy_state, training_info_ta] = tf.while_loop(
+                cond=lambda counter, *_: tf.less(counter, self._train_interval),
+                body=self._train_loop_body,
+                loop_vars=[counter, time_step, policy_state, training_info_ta],
+                back_prop=True,
+                parallel_iterations=1,
+                name='iter_loop')
+
+            training_info = tf.nest.map_structure(lambda ta: ta.stack(),
+                                                  training_info_ta)
+
+            action_distribution = nested_distributions_from_specs(
+                self._algorithm.action_distribution_spec,
+                training_info.action_distribution)
+
+            training_info = training_info._replace(
+                action_distribution=action_distribution)
+
+        if self._final_step_mode == OnPolicyDriver.FINAL_STEP_SKIP:
+            next_time_step, policy_step, _ = self._step(
+                time_step, policy_state)
+            next_state = policy_step.state
```

Comment:
why not perform final step with new GradientType and pass the new tape to next _iter and final_step_mode can be removed ?

Reply:
loop_vars only support numpy array, `Tensor`, and `TensorArray` objects

Reply:
You said it. GradientType cannot be loop_vars